tower-util/src/combinators/map_err.rs

```diff
@@ -0,0 +1,107 @@
+use futures_util::{future::MapErr as MapErrFut, TryFutureExt};
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+#[derive(Debug)]
+pub struct MapErr<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> MapErr<S, F> {
+    pub fn new(inner: S, f: F) -> Self {
+        MapErr {
+            f,
+            inner,
+        }
+    }
+}
+
+impl<S, F, Request, Error> Service<Request> for MapErr<S, F>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Error) -> Error,
+    F: Clone,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = MapErrFut<S::Future, F>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(self.f.clone())
+    }
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        self.inner.call(request).map_err(self.f.clone())
+    }
+}
+
+#[derive(Debug)]
+pub struct MapErrLayer<F, Request, Error> {
+    f: F,
+    request: PhantomData<Request>,
+    error: PhantomData<Error>,
+}
+
+impl<F, Request, Error> MapErrLayer<F, Request, Error>
+where
+    F: Clone,
+{
+    pub fn new(f: F) -> Self {
+        MapErrLayer {
+            f,
+            request: PhantomData::default(),
+            error: PhantomData::default(),
+        }
+    }
+}
+
+impl<S, F, Request, Error> Layer<S> for MapErrLayer<F, Request, Error>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Error) -> Error,
+    F: Clone,
+{
+    type Service = MapErr<S, F>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        MapErr {
+            f: self.f.clone(),
+            inner,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use tower_layer::Layer;
+
+    struct MockService;
+
+    impl Service<String> for MockService {
+        type Response = String;
+        type Error = u8;
+        type Future = futures_util::future::Ready<Result<String, u8>>;
+
+        fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+            Poll::Ready(Ok(()))
+        }
+
+        fn call(&mut self, request: String) -> Self::Future {
+            futures_util::future::ready(Ok(request))
+        }
+    }
+
+    #[test]
```

Comment:
This should be removed later. 

The purpose of these tests are to demonstrate how the API behaves.

Reply:
Could these tests be made doc comments?

Reply:
I've just added doc tests for `map_ok`, `map_err`, and `with`. Would you like more on the underlying services/layers too?

Reply:
Thanks for adding them! No, I think this is fine for now.

(Tower lacks good documentation, so we've got to start somewhere...)